### PR TITLE
Handle Pausing in AchievementManager

### PR DESCRIPTION
**Pull request by LillyJadeKatrin**
There are two pieces of functionality to be added here. One, we want to disallow pausing too frequently, as it may be used as an artificial slowdown. This is handled within the client, which can tell us if a pause is allowed. Two, we want to call rc_client_idle on a periodic basis so the connection with the server can be maintained even while the emulator is paused.

### CODE REVIEW

File: Source/Core/Core/Core.cpp [hunk 1]
```diff
@@ -707,11 +707,18 @@ void SetState(Core::System& system, State state, bool report_state_change)
   switch (state)
   {
   case State::Paused:
+#ifdef USE_RETRO_ACHIEVEMENTS
+    if (!AchievementManager::GetInstance().CanPause())
+      return;
```
Comment (JosJuice) [inline Source/Core/Core/Core.cpp:712]: You should skip this check when called from CPUSetInitialExecutionState. Otherwise the state might stay as Starting when it shouldn't.

File: Source/Core/Core/AchievementManager.cpp [hunk 2]
```diff
@@ -239,6 +239,37 @@ void AchievementManager::DoFrame()
...
+  if (!IsGameLoaded())
+    return;
+
+  m_idle_thread = new std::thread([]() {
+    auto* client = AchievementManager::GetInstance().m_client;
+    auto* system = AchievementManager::GetInstance().m_system;
+    while (Core::GetState(*system) == Core::State::Paused)
+    {
+      rc_client_idle(client);
+      Common::SleepCurrentThread(1000);
```
Comment (JosJuice) [inline Source/Core/Core/AchievementManager.cpp:268]: This is bad for battery life on mobile platforms, since there's nothing preventing the loop from continuing when the screen is off or another app is in the foreground. While there's no way of enabling RetroAchievements in the Android GUI right now, if you want to add a loop like this, I request that you add an explicit check to not activate RetroAchivements when `ANDROID` is defined.

**Review by JosJuice — COMMENTED** [review]
You should also audit all calls to EnableStepping. For instance, right now I don't think there's anything stopping the GDB stub from pausing emulation.

File: Source/Core/Core/AchievementManager.cpp [hunk 3]
```diff
@@ -239,6 +239,37 @@ void AchievementManager::DoFrame()
...
+  }
+  return can_pause;
+}
+
+void AchievementManager::DoIdle()
+{
+  if (!IsGameLoaded())
+    return;
+
+  m_idle_thread = new std::thread([]() {
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/AchievementManager.cpp:262]: This is nonsense. Where do I even start here.

- This thread is never joined, so it will just live forever.
- It is also never deleted, leaking memory.
- Every call to DoIdle() will create a new thread even if one already exists.
- Nothing in the thread synchronizes with the outside world. `m_client` may become invalid at any time and you'd have no idea.

Comment (JosJuice) [reply Source/Core/Core/AchievementManager.cpp:262]: > This thread is never joined, so it will just live forever.

Doesn't it stop executing within a second or so of the state becoming Paused?

I very much agree regarding the memory leak and the validity of m_client, though. (The validity of m_system could possibly also become relevant in the future.)

File: Source/Core/Core/AchievementManager.cpp (hunk 2 above)
Comment (LillyJadeKatrin) [reply Source/Core/Core/AchievementManager.cpp:262]: Once https://github.com/JosJuice/dolphin/commit/adc0e051ae64d6c617a12b93f44b9dc1e003400a is merged I'll be using that to only run the idle loop when the app is foregrounded.

Comment (JosJuice) [reply Source/Core/Core/AchievementManager.cpp:262]: I'm not planning to create any separate PR for that commit, as it has no use on its own. Please cherry-pick it either into this PR or into a later follow-up PR.

File: Source/Core/Core/AchievementManager.cpp [hunk 4]
```diff
@@ -239,6 +239,37 @@ void AchievementManager::DoFrame()
...
+  return can_pause;
+}
+
+void AchievementManager::DoIdle()
+{
+  m_idle_thread = std::make_unique<std::thread>([]() {
+    while (Core::GetState(*AchievementManager::GetInstance().m_system) == Core::State::Paused)
+    {
+      Common::SleepCurrentThread(1000);
+      std::lock_guard lg{AchievementManager::GetInstance().GetLock()};
```
Comment (JosJuice) [inline Source/Core/Core/AchievementManager.cpp:263]: The other thread isn't locking this lock when reassigning m_client.

Comment (JosJuice) [reply Source/Core/Core/AchievementManager.cpp:263]: I also don't recommend using the same lock for every piece of state that needs locking, as this increases lock contention. If you need to protect `m_client`, you can have a lock for just `m_client`.

**Comment by LillyJadeKatrin** [comment]
> You should also audit all calls to EnableStepping. For instance, right now I don't think there's anything stopping the GDB stub from pausing emulation.

I looked over EnableStepping and the GDB stub does appear to be the only outlier. I'll disable that in hardcore mode in a separate PR.

File: Source/Core/Core/AchievementManager.cpp (hunk 2 above)
Comment (JosJuice) [reply Source/Core/Core/AchievementManager.cpp:262]: And for clarity, if you want to address the Android problem in a follow-up PR, please explicitly disable RetroAchivements on Android for now.

File: Source/Core/Core/AchievementManager.cpp [hunk 5]
```diff
@@ -239,6 +249,38 @@ void AchievementManager::DoFrame()
...
+        OSD::Duration::VERY_LONG, OSD::Color::RED);
+  }
+  return can_pause;
+}
+
+void AchievementManager::DoIdle()
+{
+  m_idle_thread = std::make_unique<std::thread>([]() {
+    while (Core::GetState(*AchievementManager::GetInstance().m_system) == Core::State::Paused &&
+           AchievementManager::GetInstance().m_background_execution_allowed)
```
Comment (JosJuice) [inline Source/Core/Core/AchievementManager.cpp:271]: The way you're using m_background_execution_allowed, it technically needs to be an atomic to avoid undefined behavior.

File: Source/Core/Core/AchievementManager.cpp [hunk 6]
```diff
@@ -239,6 +249,38 @@ void AchievementManager::DoFrame()
...
+    {
+      Common::SleepCurrentThread(1000);
+      std::lock_guard lg{AchievementManager::GetInstance().m_client_lock};
+      auto* client = AchievementManager::GetInstance().m_client;
+      if (!client || !AchievementManager::GetInstance().IsGameLoaded())
+        return;
+      rc_client_idle(client);
```
Comment (JosJuice) [inline Source/Core/Core/AchievementManager.cpp:278]: `m_client_lock` is guarding against the other thread reassigning `m_client`, but not against the other thread using the struct pointed to by `m_client`. So I need to ask: Is calling `rc_client_idle` while the other thread is also using `m_client` thread-safe? The rcheevos documentation for `rc_client_idle` doesn't say anything about it as far as I can tell. There is a mutex being locked inside `rc_client_idle`, but I don't know if that means the whole function is safe or not.

Comment (LillyJadeKatrin) [reply Source/Core/Core/AchievementManager.cpp:278]: I've gone over rc_client_idle and my end takeaway is that it is _not_ threadsafe enough for this. As a result, I would prefer to lock every use of m_client, and can use the original m_lock to do so because it's already everywhere.

**Comment by JosJuice** [comment]
I don't think I can see any issues anymore, but I would like @AdmiralCurtiss to take a look at this too.

File: Source/Core/Core/AchievementManager.cpp [hunk 7]
```diff
@@ -239,6 +249,38 @@ void AchievementManager::DoFrame()
...
+    {
+      Common::SleepCurrentThread(1000);
+      std::lock_guard lg{AchievementManager::GetInstance().m_lock};
+      auto* client = AchievementManager::GetInstance().m_client;
+      if (!client || !AchievementManager::GetInstance().IsGameLoaded())
+        return;
+      rc_client_idle(client);
+    }
+  });
+  m_idle_thread->detach();
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/AchievementManager.cpp:281]: There's still something iffy here. If you detach the thread immediately why store it in a member variable at all?

I also generally think that casting a thread into the ether and hoping it synchronizes itself with regular polls is a poor idea (you cannot guarantee that this ever finishes, because in theory every time it polls the core state might be paused at this very moment, even if it was non-paused between the polls), but I suppose it's *okay enough* to work in practice and we can always refactor it later.

File: Source/Core/Core/AchievementManager.h [hunk 8]
```diff
@@ -203,6 +208,7 @@ class AchievementManager
   std::unordered_map<AchievementId, Badge> m_locked_badges;
   RichPresence m_rich_presence;
   std::chrono::steady_clock::time_point m_last_rp_time = std::chrono::steady_clock::now();
+  std::unique_ptr<std::thread> m_idle_thread = nullptr;
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/AchievementManager.h:211]: this is unused now

File: Source/Core/Core/AchievementManager.cpp [hunk 9]
```diff
@@ -246,6 +252,37 @@ void AchievementManager::DoFrame()
...
+        fmt::format("Can pause in {} seconds.", static_cast<float>(frames_to_next_pause) / 60),
+        OSD::Duration::VERY_LONG, OSD::Color::RED);
+  }
+  return can_pause;
+}
+
+void AchievementManager::DoIdle()
+{
+  std::thread([]() {
+    while (Core::GetState(*AchievementManager::GetInstance().m_system) == Core::State::Paused &&
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/AchievementManager.cpp:273]: I played around with this a little bit and crashed. Turns out `m_system` can be nullptr here! Which makes sense if you look at `CloseGame()`. May I suggest changing this loop to:

```
  std::thread([this]() {
    while (true)
    {
      Common::SleepCurrentThread(1000);
      std::lock_guard lg{m_lock};
      if (!m_system || Core::GetState(*m_system) != Core::State::Paused)
        return;
      if (!m_background_execution_allowed)
        return;
      if (!m_client || !IsGameLoaded())
        return;
      rc_client_idle(m_client);
    }
  }).detach();
```

**Comment by LillyJadeKatrin** [comment]
And I got an assert failure because apparently do_idle is processing rich presence despite that not actually updating at all while paused. Looking into it.

**Comment by CasualPokePlayer** [comment]
Why not just have this thread constantly call Core::QueueHostJob? That would ensure it gets used constantly can idle can end up doing memory reads just fine (and it's probably better to touch cheevo stuff on the host thread rather)

File: Source/Core/Core/AchievementManager.cpp [hunk 10]
```diff
@@ -242,6 +252,41 @@ void AchievementManager::DoFrame()
...
+      std::lock_guard lg{m_lock};
+      if (!m_system || Core::GetState(*m_system) != Core::State::Paused)
+        return;
+      if (!m_background_execution_allowed)
+        return;
+      Core::QueueHostJob([this](Core::System& system) {
+        if (!m_client || !IsGameLoaded())
+          return;
+        rc_client_idle(m_client);
+      });
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/AchievementManager.cpp:285]: While I do agree that QueueHostJob is probably not a bad idea, this is not how you do it. Think about the threading implications here: Your lock from the local thread does not propagate to the host job. The `return;` here will also now no longer stop the local thread, because it just exits the inner lambda.

**Comment by LillyJadeKatrin** [comment]
Still evaluating but ran out of time this morning so here's what it looks like so far.

**Comment by AdmiralCurtiss** [comment]
So are you done with this now?

**Comment by LillyJadeKatrin** [comment]
I'm not finding anything wrong with it yet, I think I might have this locked in.

**Comment by AdmiralCurtiss** [comment]
Okay, I'll give this another test then. I still think the unsynchronized thread is fundamentally flawed, but whatever, we can go with it for now.

File: Source/Core/Core/AchievementManager.cpp [hunk 11]
```diff
@@ -457,6 +511,7 @@ void AchievementManager::Shutdown()
   if (m_client)
   {
     CloseGame();
+    std::lock_guard lg{m_lock};
```
Comment (AdmiralCurtiss) [inline Source/Core/Core/AchievementManager.cpp:514]: Locking here can lead to a deadlock. You can't know how many events the queue will still process after you lock this but before it processes the Shutdown request, which can lead to the situation where the queue waits for the lock to become free and this thread waits for the queue to shut down. Move the lock after the Shutdown().

File: Source/Core/Core/AchievementManager.cpp [hunk 12]
```diff
@@ -246,6 +252,50 @@ void AchievementManager::DoFrame()
...
+bool AchievementManager::CanPause()
+{
+  u32 frames_to_next_pause = 0;
+  bool can_pause = rc_client_can_pause(m_client, &frames_to_next_pause);
+  if (!can_pause)
+  {
+    OSD::AddMessage("Cannot spam pausing in hardcore mode.", OSD::Duration::VERY_LONG,
+                    OSD::Color::RED);
+    OSD::AddMessage(
+        fmt::format("Can pause in {} seconds.", static_cast<float>(frames_to_next_pause) / 60),
```
Comment (JosJuice) [inline Source/Core/Core/AchievementManager.cpp:264]: This shouldn't be using a hardcoded 60 (unless RetroAchievements has a hardcoded behavior of considering 16.67 ms to be one frame, which I really hope it doesn't). You can get the actual refresh rate from GetTargetRefreshRate in VideoInterface.h. You may also want to adjust for the emulation speed, though I'm not how sure how useful that is since you can start and stop fast forwarding very quickly, so I'll leave the choice on that one to you.

(Actually, how does rcheevos know the current framerate? Is it just not taking the framerate into account for determining when you can pause, making the pause interval longer for 50 Hz games in terms of seconds?)

File: Source/Core/Core/AchievementManager.cpp [hunk 13]
```diff
@@ -246,6 +252,50 @@ void AchievementManager::DoFrame()
...
+      {
+        std::lock_guard lg{m_lock};
+        if (!m_system || Core::GetState(*m_system) != Core::State::Paused)
+          return;
+        if (!m_background_execution_allowed)
+          return;
+        if (!m_client || !IsGameLoaded())
+          return;
+      }
+      Core::QueueHostJob([this](Core::System& system) {
```
Comment (JosJuice) [inline Source/Core/Core/AchievementManager.cpp:285]: Would be nice to have a comment in this function about why we're moving this to the host thread. Weird workarounds like this are the kind of thing that are in most need of comments, as the comments help people in the future understand not only why we have this code to begin with but also whether they can safely remove it.

File: Source/Core/Core/AchievementManager.cpp (hunk 12 above)
Comment (LillyJadeKatrin) [reply Source/Core/Core/AchievementManager.cpp:264]: It assumes 60Hz. No, the logic behind it doesn't make much sense to me either.

**Comment by AdmiralCurtiss** [comment]
I'm still not 100% happy with this but I think we can go with this for now...

**Comment by DevanWolf** [comment]
DO NOT USE THIS VERSION OF THE APP!
It has a flaw which thinks every game is banned.

**Comment by kira1000** [comment]
> DO NOT USE THIS VERSION OF THE APP! It has a flaw which thinks every game is banned.

where does it say that?  knowing would make it easier to verify.

**Comment by DevanWolf** [comment]
> > DO NOT USE THIS VERSION OF THE APP! It has a flaw which thinks every game is banned.
> 
> where does it say that? Are you sure that's what you mean and not that there are no achievements?

No, the version is broken, launch any game and it crashes.

**Comment by LillyJadeKatrin** [comment]
> > > DO NOT USE THIS VERSION OF THE APP! It has a flaw which thinks every game is banned.
> > 
> > 
> > where does it say that? Are you sure that's what you mean and not that there are no achievements?
> 
> No, the version is broken, launch any game and it crashes.

We have a team of people actively testing and verifying this integration and the only thing remotely similar to the behavior you describe is a bug on the Android version which has since been fixed. I invite you to join the Dolphin Discord server if you continue to have issues but here is a terrible place to discuss them.

**Comment by JosJuice** [comment]
Or the forums, or the IRC channel, or the issue tracker.
